api/valor_api/backend/metrics/llm_call.py

```diff
@@ -0,0 +1,277 @@
+from typing import Any
+
+import openai
+from mistralai.client import MistralAPIException, MistralClient
+from mistralai.models.chat_completion import ChatMessage
+
+COHERENCE_INSTRUCTION = """You are a helpful assistant. You will grade the user's text. Your task is to rate the text based on its coherence. Please make sure you read and understand these instructions carefully. Please keep this document open while reviewing, and refer to it as needed.
+Evaluation Criteria:
+Coherence (1-5) - the collective quality of all sentences. We align this dimension with the DUC quality question of structure and coherence whereby ”the summary should be well-structured and well-organized. The summary should not just be a heap of related information, but should build from sentence to sentence to a coherent body of information about a topic.”
+Evaluation Steps:
+1. Read the text carefully and identify the main topic and key points.
+2. Assign a score for coherence on a scale of 1 to 5, where 1 is the lowest and 5 is the highest based on the Evaluation Criteria. Respond with just the number 1 to 5."""
```

Comment:
I should cite Deepeval for this instruction. See appendix A of https://arxiv.org/pdf/2303.16634 for where I got this instruction from. 

However I also significantly modified their instruction. Deepeval mentions in their paper that the coherence metric should measure how coherently the output (Prediction.Annotation.text) follows from the input (Datum.text). However the definition of coherence you see here (which is exactly copied from an example in Deepeval's paper) sounds pretty different. It makes no mention of measuring the coherence of just the output, and instead defines coherence as "the collective quality of all sentences". 

The Deepeval example instruction instead uses the evaluation steps to indicate to the LLM that it should be judging the output text. However the evaluation steps in the example are overly specific (specifically designed for news article summarization). Deepeval deals with this by prompting an LLM to come up with evaluation steps for the given task (an approach they call G-Eval). We could take the same approach, however I wanted to do something simpler and easier for the user, at least for now, so I decided to simplify the evaluation steps and not include the input (Datum.text) in the LLM prompt. 

I think my approach seems better to Valor, as G-Eval seems overly involved compared to other metrics in Valor, however let me know if you think we should stick closer to Deepeval's approach.

Reply:
This is the example from Deepeval's paper

You will be given one summary written for a news article.
Your task is to rate the summary on one metric.
Please make sure you read and understand these instructions carefully. Please keep this
document open while reviewing, and refer to it as needed.
Evaluation Criteria:
Coherence (1-5) - the collective quality of all sentences. We align this dimension with
the DUC quality question of structure and coherence whereby ”the summary should be
well-structured and well-organized. The summary should not just be a heap of related information, but should build from sentence to sentence to a coherent body of information about a topic.”
Evaluation Steps:

Read the news article carefully and identify the main topic and key points.
Read the summary and compare it to the news article. Check if the summary covers the main
topic and key points of the news article, and if it presents them in a clear and logical order.
Assign a score for coherence on a scale of 1 to 5, where 1 is the lowest and 5 is the highest
based on the Evaluation Criteria.
Example:
Source Text:
{{Document}}
Summary:
{{Summary}}
Evaluation Form (scores ONLY):
Coherence:

Reply:
I modified this to a function. 

I am concerned though that I have strayed too far from the conventional use of coherence.